File: programmers/level2/배달.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define INF 987654321

vector<vector<pair<int, int>>> adjust;
vector<int> dist;

void dijkstra(int start)
{
    priority_queue<pair<int, int>> pq;
    pq.push({0, start});
    dist[start] = 0;

    while (!pq.empty())
    {
        int nowDist = -pq.top().first;
        int nowNode = pq.top().second;
        pq.pop();

        for (int i = 0; i < adjust[nowNode].size(); i++)
        {
            int cost = nowDist + adjust[nowNode][i].second;

            if (cost < dist[adjust[nowNode][i].first])
            {
                dist[adjust[nowNode][i].first] = cost;
                pq.push({-cost, adjust[nowNode][i].first});
            }
        }
    }
}

int solution(int N, vector<vector<int>> road, int K)
{
    int answer = 0;
    adjust.resize(N + 1);
    dist.resize(N + 1, INF);

    for (int i = 0; i < road.size(); i++)
    {
        int a = road[i][0];
        int b = road[i][1];
        int c = road[i][2];

        adjust[a].push_back({b, c});
        adjust[b].push_back({a, c});
    }

    dijkstra(1);

    for (int i = 1; i <= N; i++)
    {
        if (dist[i] <= K)
            answer++;
    }

    return answer;
}

int main() {}

File: programmers/level2/배달.py
from queue import PriorityQueue

INF = 987654321


def dijsktra(dist, adjust, K):

    pq = PriorityQueue()
    pq.put((0, 1))
    dist[1] = 0

    while not pq.empty():
        now = pq.get()
        now_dist = -now[0]
        now_node = now[1]

        for next in adjust[now_node]:
            cost = now_dist + next[1]

            if cost < dist[next[0]]:
                dist[next[0]] = cost
                pq.put((-cost, next[0]))


def solution(N, roads, K):

    dist = [INF for _ in range(N+1)]
    adjust = [[] for _ in range(N+1)]

    for road in roads:
        s = road[0]
        e = road[1]
        w = road[2]
        adjust[s].append((e, w))
        adjust[e].append((s, w))

    dijsktra(dist, adjust, K)

    answer = 0

    for i in range(1, N+1):
        if dist[i] <= K:
            answer += 1

    return answer
